File: src/main/java/app/askresume/global/interceptor/AdminAuthorizationInterceptor.java
package app.askresume.global.interceptor;

import app.askresume.domain.member.constant.Role;
import app.askresume.global.jwt.service.TokenManager;
import app.askresume.global.error.ErrorCode;
import app.askresume.global.error.exception.AuthenticationException;
import io.jsonwebtoken.Claims;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

@Component
@RequiredArgsConstructor
public class AdminAuthorizationInterceptor implements HandlerInterceptor {

    private final TokenManager tokenManager;

    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {

        String authorizationHeader = request.getHeader(HttpHeaders.AUTHORIZATION);
        String accessToken = authorizationHeader.split(" ")[1];

        Claims tokenClaims = tokenManager.getTokenClaims(accessToken);
        String role = (String)tokenClaims.get("role");
        if(!Role.ADMIN.equals(Role.valueOf(role))) {
            throw new AuthenticationException(ErrorCode.FORBIDDEN_ADMIN);
        }

        return true;
    }

}

File: src/main/kotlin/app/askresume/global/interceptor/AdminAuthorizationInterceptor.kt
package app.askresume.global.interceptor

import app.askresume.domain.member.constant.Role
import app.askresume.global.error.exception.ForbiddenAdminException
import app.askresume.global.jwt.service.TokenManager
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class AdminAuthorizationInterceptor(
    private val tokenManager: TokenManager,
    @Value("\${spring.profiles.active}") var profile: String
) : HandlerInterceptor {

    @Throws(Exception::class)
    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        if (request.method.uppercase() == HttpMethod.OPTIONS.name.uppercase()) return true
        if (profile.uppercase() == PROFILE_LOCAL) return true

        val authorizationHeader = request.getHeader(HttpHeaders.AUTHORIZATION)
        val accessToken = authorizationHeader.split(" ")[1]

        val tokenClaims = tokenManager.getTokenClaims(accessToken)
        val role = tokenClaims["role"] as String
        if (Role.ADMIN != Role.valueOf(role)) {
            throw ForbiddenAdminException()
        }

        return true
    }

    companion object {
        private const val PROFILE_LOCAL = "LOCAL"
    }
}
